### ISSUE TITLE
Enforcing Connection.setReadOnly

### ISSUE THREAD

**Issue by christopheblin** [jdbi/jdbi#1241]
I have a use case which I think is very common : in dev, my db is a single instance whereas in prod it is a master-slaves cluster

So I am using java.sql.Connection.setReadOnly when I open a connection (using true when the business is readonly)

The problem I have faced is that everything works in dev but then breaks in prod with the following error : org.jdbi.v3.core.statement.UnableToExecuteStatementException: java.sql.SQLException: (conn=37313) The MySQL server is running with the --read-only option so it cannot execute this statement 

This happens because someone has inadvertently modified the code without changing the setReaonly flag properly.

I was wondering if there is something I can do to throw an exception in dev

I am using SqlObjects in kotlin (i.e @SqlUpdate to perform an insert/update/delete)

basically, what I would like is to modify Handle.java / createUpdate to throw an error if handle has been set to readonly == true.

However, Handle is NOT an interface (so I can not use decorator pattern for example). Do you know if there is another solution to achieve what I want ?

**Comment by qualidafial** [comment]
I recommend you spin up two Jdbi instances: one with a read/write DataSource, and one with a read-only DataSource.

In your app, you'll need to make sure you use the right Jdbi instance depending on whether you're reading or writing.

**Comment by qualidafial** [comment]
So it might be possible to make `BaseStatement.customizers` backed by a config property, rather than independent per statement. That way you can call e.g. `jdbi.getConfig(SqlStatements.class).addCustomizer(new ReadOnlyCheckCustomizer())` to configure a statement customizer globally. Would that help?

**Comment by christopheblin** [comment]
@qualidafial thanks for the answers but I am sorry I do not understand fully what you mean

I am already able to set readonly flag on the connection deending on whereas I need to read or write but maybe this was not clear in my description (or maybe you mean something else ?).

Second, I dont get what I can achieve with jdbi.getConfig(SqlStatements.class).addCustomizer(new ReadOnlyCheckCustomizer())
Does it mean I must set this on the readonly instance ? should I do something inside this customizer to detect if the current handle is readonly ?

I will try to inestigate ow to use customizers since I ma not familiar with that concept at all....
If you can give me additional pointers, do not hesitate :)

**Comment by qualidafial** [comment]
Jdbi provides hooks to run custom code before or after executing a statement. http://jdbi.org/apidocs/org/jdbi/v3/core/statement/StatementCustomizer.html

You could write an implementation of this interface (e.g. `ReadOnlyCheckCustomizer` and have it throw an exception if the connection is read-only and the statement type is an `Update`.

Currently it is only possible to register a statement customizer on individual statements e.g. `handle.createUpdate(sql).addCustomizer(new ReadOnlyCheckCustomizer()).bind(...).execute()`. However, we could make it possible to register global customers, so it would apply to all statements.

**Comment by christopheblin** [comment]
@qualidafial I have investigate again after you gave me the pointer, and I now understand better the concept of customizers

Would it be possible to create an annotation @ReadOnlyCheck like @QueryTimeout ?

For sure, the global customizer would be less work for the devs because it will apply automatically so I am still open to that solution if it comes in a near release

**Comment by christopheblin** [comment]
Following instruction at http://jdbi.org/#_statement_customizing_annotations, here is my solution for SQL objects until there is a way to perform global customizers

    package org.jdbi.v3.sqlobject.customizer;
    
    import org.jdbi.v3.sqlobject.customizer.internal.ReadOnlyCheckFactory;
    
    import java.lang.annotation.ElementType;
    import java.lang.annotation.Retention;
    import java.lang.annotation.RetentionPolicy;
    import java.lang.annotation.Target;
    
    @Retention(RetentionPolicy.RUNTIME)
    @Target({ElementType.METHOD})
    @SqlStatementCustomizingAnnotation(ReadOnlyCheckFactory.class)
    public @interface ReadOnlyCheck {
    }


    package org.jdbi.v3.sqlobject.customizer.internal;
    
    import org.jdbi.v3.sqlobject.customizer.SqlStatementCustomizer;
    import org.jdbi.v3.sqlobject.customizer.SqlStatementCustomizerFactory;
    
    import java.lang.annotation.Annotation;
    import java.lang.reflect.Method;
    
    public class ReadOnlyCheckFactory implements SqlStatementCustomizerFactory {
        @Override
        public SqlStatementCustomizer createForMethod(Annotation annotation, Class<?> sqlObjectType, Method method) {
            return stmt -> stmt.addCustomizer(new ReadOnlyCheckCustomizer());
        }
    }


    package org.jdbi.v3.sqlobject.customizer.internal;
    
    import org.jdbi.v3.core.statement.StatementContext;
    import org.jdbi.v3.core.statement.StatementCustomizer;
    
    import java.sql.PreparedStatement;
    import java.sql.SQLException;
    
    public class ReadOnlyCheckCustomizer implements StatementCustomizer {
    
        public static final String EXCEPTION = "can not perform this operation on a readonly connection";
    
        @Override
        public void beforeExecution(PreparedStatement stmt, StatementContext ctx) throws SQLException {
            if (stmt.getConnection().isReadOnly()) {
                throw new SQLException(EXCEPTION);
            }
        }
    }

**Comment by christopheblin** [comment]
And this can be used like this :

    interface DaoForTest {
        @SqlUpdate("update mytable set id = '' where 1 = 0")
        @ReadOnlyCheck
        fun test()
    }

**Comment by qualidafial** [comment]
Bingo.

I've opened #1245 to discuss global statement customizer config.

**Comment by christopheblin** [comment]
Thanks for the help, I close this issue
